q2_feature_classifier/_consensus_assignment.py

```diff
@@ -0,0 +1,252 @@
+        if _id not in obs_taxa:
+            obs_taxa[_id] = (unassignable_label, 0.0)
+
+    return obs_taxa
+
+
+# Replace this function with QIIME2 API for wrapping commands/binaries,
+# pending https://github.com/qiime2/qiime2/issues/224
+def _run_command(cmd, verbose=True):
+    if verbose:
+        print("Running external command line application. This may print "
+              "messages to stdout and/or stderr.")
+        print("The command being run is below. This command cannot "
+              "be manually re-run as it will depend on temporary files that "
+              "no longer exist.")
+        print("\nCommand:", end=' ')
+        print(" ".join(cmd), end='\n\n')
+    subprocess.run(cmd, check=True)
+
+
+def _import_blast_assignments(
+        assignments, ref_taxa,
+        unassignable_label=_get_default_unassignable_label()):
+    '''import observed assignments in blast6 format to dict of lists.
+
+    assignments: path or list
+        Taxonomy observation map in blast format 6 or 7. Each line consists of
+        taxonomy assignments of a query sequence in tab-delimited format:
+            <query_id>    <assignment_id>   <...other columns are ignored>
+
+    ref_taxa: dict or pd.Series
+        Reference taxonomies in tab-delimited format:
+            <accession ID>  kingdom;phylum;class;order;family;genus;species
+    '''
+    obs_taxa = {}
+    lines = _open_list_or_file(assignments)
+
+    for line in lines:
+        if not line.startswith('#') or line == "":
+            fields = line.split('\t')
+            # if vsearch fails to find assignment, it reports '*' as the
+            # accession ID, which is completely useless and unproductive.
+            if fields[1] == '*':
+                t = [unassignable_label]
+            else:
+                # ref taxonomy IDs get imported as a str when using CLI
+                # but imported as int in python interpreter (e.g., for tests)
+                # the following allows dict lookup as str or int.
+                if ref_taxa.index.dtype == int:
+                    id_ = int(fields[1])
+                else:
+                    id_ = fields[1]
+
+                try:
+                    t = ref_taxa[id_]
+                except KeyError:
+                    raise KeyError((
+                        'Identifier {0} was reported in taxonomic search '
+                        'results, but was not present in the reference '
+                        'taxonomy.').format(str(id_)))
+
+                try:
+                    t = t.split(';')
+                except ValueError:
+                    raise ValueError((
+                        'Reference taxonomy {0} (id: {1}) is incorrectly '
+                        'formatted.').format(t, str(id_)))
+
+            if fields[0] in obs_taxa.keys():
+                obs_taxa[fields[0]].append(t)
+            else:
+                obs_taxa[fields[0]] = [t]
+    return obs_taxa
+
+
+def _open_list_or_file(infile):
```

Comment:
Yes. The assignments are first output to a temp file by the blast/vsearch subprocess. This function is used to import this to a list — but is also convenient for unit testing, where I can test this function without importing a data file, hence my desire for flexibility here.